### PR TITLE
pk-engine: Always allow user interaction for offline update/upgrade

### PR DESCRIPTION
**Pull request by mcrha**
This partly reverts commit 9b7e083cf849c4ed4d66fe32250f1615ab577d94.
The offline update/upgrade triggers do need the user interaction, to
get the credentials from the polkit, otherwise the trigger cannot
do its job.

Closes https://github.com/PackageKit/PackageKit/issues/504

### CODE REVIEW

**Comment by hughsie** [comment]
@pwithnall please

**Review by pwithnall — CHANGES_REQUESTED** [review]
I don’t think this is the right fix. Whether to allow user interaction should only ever be set at the top level of a call stack, typically where some code is being called from the UI or from a background service. Only then is it possible to know whether user interaction is definitely allowed/not allowed.

On the basis of what’s written in #504 I think the right fix would be to add an `interactive` property to the relevant bit of `PkOffline`.

**Comment by mcrha** [comment]
The only problem is that the PkOffline is not an object, it's a call wrapper with bare functions. It's doable, of course.

**Comment by lruzicka** [comment]
I would like to add that this issue is an accepted Beta blocker issue on Fedora 35, so it should be resolved possibly without much hesitation. If the correct fix takes much time to do, maybe the suggested fix could be used over the Beta release and revisited afterwards?

**Comment by mcrha** [comment]
I just updated the branch to follow what the #504 suggests as a better solution.

File: lib/packagekit-glib2/pk-offline.c
Comment (pwithnall) [inline lib/packagekit-glib2/pk-offline.c:192]:
```suggestion
 * Clears the last offline operation report, which may be success or failure.
```

File: lib/packagekit-glib2/pk-offline.h
Comment (pwithnall) [inline lib/packagekit-glib2/pk-offline.h:77]:
```suggestion
 * @PK_OFFLINE_FLAGS_INTERACTIVE:	Run the action in an interactive mode, allowing polkit authentication dialogs
```

**Review by pwithnall — COMMENTED** [review]
Two really minor nitpicks, but otherwise this looks great to me (I’m not a maintainer though)

File: lib/packagekit-glib2/pk-offline.c
Comment (mcrha) [reply lib/packagekit-glib2/pk-offline.c:192]: Copy&paste thing, the original function has it too ;)

**Comment by mcrha** [comment]
I squashed your suggestions to the main commit.

**Review by hughsie — APPROVED** [review]
Great stuff, thanks.

**Comment by mcrha** [comment]
One thing I forgot to mention. The old pk_offline() functions call as non-interactive. Would it make sense to rather use the interactive mode for those always, as it's better to ask the credentials with the old API than to "always" fail?

**Comment by hughsie** [comment]
> as it's better to ask the credentials with the old API than to "always" fail?

So "working more than before" or "behaviour change". I don't mind either way. I guess I'd prefer it worked in more cases, but I'm worried about things like openQA changing. I'll defer to your judgement here.

**Comment by lruzicka** [comment]
> > as it's better to ask the credentials with the old API than to "always" fail?
> 
> So "working more than before" or "behaviour change". I don't mind either way. I guess I'd prefer it worked in more cases, but I'm worried about things like openQA changing. I'll defer to your judgement here.

As for OpenQA, I guess we will have to deal with changes, but we'd prefer if the behaviour was somehow consistent and we did know what happens under certain conditions :D

**Comment by pwithnall** [comment]
I think it’s probably safest to have the default be to behave the same as before https://github.com/PackageKit/PackageKit/commit/9b7e083cf849c4ed4d66fe32250f1615ab577d94, since that was quite a recent change. That would maintain backwards compatibility for callers. So that means allowing user interaction by default.

**Comment by mcrha** [comment]
I agree it makes sense to allow the user interaction for the "old" functions, that's what it used to be until recently. I updated the patch accordingly.

**Pull request merged**
